README.md
This is an HTTP frontend for the validator.nu HTML 5 parser. It parses some 
input text and returns the reserialized HTML.

## Compile and test

Ubuntu build/test dependencies:
* openjdk-7-jdk
* maven2
* jsvc

Compile and generate a .jar file by running `mvn package`.

We use the "shade" plugin to bundle all dependencies. So to run it, you only
need the generated .jar file in the class path. To start the server as a
daemon, use something like:

```
/usr/bin/jsvc \
	-cp $(pwd)/target/html5depurate-1.0-SNAPSHOT.jar \
	-pidfile /tmp/html5depurate.pid \
	-errfile /tmp/html5depurate.err \
	-outfile /tmp/html5depurate.out \
	-procname html5depurate \
	org.wikimedia.html5depurate.DepurateDaemon
```

The default log format is pretty bad but can be configured by the usual means,
with -Djava.util.logging.config.file=/path/to/logging.properties

Then to test:

```
curl http://localhost:4339/document -F text=foo
```

This will return an HTML document which is a reserialized version of "foo".

## Configuration

Configuration options may be specified in /etc/html5depurate/html5depurate.conf.
Possible configuration options and their default values are documented below:

```
# Max POST size, in bytes.
maxPostSize = 100000000

# Host or IP and port on which Html5depurate will listen.
host = localhost
port = 4339
```

## To do

* Debian packaging
  - A SysV init script wrapping jsvc should be fairly simple.
  - Very strong security guarantees are possible by using a security.policy
    file.
  - There is no package for grizzly, so we will have to bundle it for now.
    Using Maven Central during build, instead of creating about 9 new Debian
    source packages, is not allowed in Debian upstream, but WMF can distribute
    the resulting file.

* Collect warnings/errors and provide a JSON serialized return format
  exposed at /info.

* Help out MW a bit by extracting the contents of the body tag. This could be
  provided at /body.

* A servlet version, if someone needs that. An early version depended on a
  servlet container, but I abandoned that approach in favour of the robustness
  and management simplicity of a standalone daemon.


debian/changelog
html5depurate (1.1.5~trusty) unstable; urgency=medium

  * Backport to trusty

 -- Tim Starling <tstarling@wikimedia.org>  Tue, 06 Dec 2016 09:19:08 +1100

html5depurate (1.1.5) unstable; urgency=medium

  * Fix bug with incorrect reset of output streams during tag stack split
    (patch by Subramanya Sastry)

 -- Tim Starling <tstarling@wikimedia.org>  Tue, 06 Dec 2016 09:15:27 +1100

html5depurate (1.1.4~trusty) unstable; urgency=medium

  * Backport to trusty

 -- Tim Starling <tstarling@wikimedia.org>  Fri, 02 Dec 2016 10:27:50 +1100

html5depurate (1.1.4) unstable; urgency=medium

  * Eliminate edge cases from p-wrapping and block tag interaction (patch by
    Subramanya Sastry).

 -- Tim Starling <tstarling@wikimedia.org>  Fri, 02 Dec 2016 10:23:21 +1100

html5depurate (1.1.3~trusty) trusty; urgency=medium

  * Backport to trusty

 -- Tim Starling <tstarling@wikimedia.org>  Fri, 11 Nov 2016 14:32:14 +1100

html5depurate (1.1.3) unstable; urgency=medium

  * Fix p-wrapping of non-phrasing content (T150317)
  * Fix Maven build for non-writable $HOME under pbuilder
    https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=441052

 -- Tim Starling <tstarling@wikimedia.org>  Fri, 11 Nov 2016 14:26:14 +1100

html5depurate (1.1.2~trusty) trusty; urgency=medium

  * Backport to trusty

 -- Tim Starling <tstarling@wikimedia.org>  Fri, 22 Jul 2016 16:30:30 +1000

html5depurate (1.1.2) unstable; urgency=medium

  * Fix jar file installed location per Debian policy
  * Limit log file size

 -- Tim Starling <tstarling@wikimedia.org>  Fri, 22 Jul 2016 15:43:04 +1000

html5depurate (1.1.1) unstable; urgency=medium

  * Add p to the list of marked empty elements

 -- Tim Starling <tstarling@wikimedia.org>  Wed, 25 May 2016 13:21:41 +1000

html5depurate (1.1.0) unstable; urgency=medium

  * Fix p-wrapping bugs

 -- Tim Starling <tstarling@wikimedia.org>  Tue, 17 May 2016 13:17:26 +1000

html5depurate (1.0.2~trusty) stable; urgency=medium

  * Backport to trusty

 -- Tim Starling <tstarling@wikimedia.org>  Fri, 29 Apr 2016 10:31:28 +1000
html5depurate (1.0.2) unstable; urgency=medium

  * Fix jdk/jre dependencies. The init script relies on default-jre being
    installed.

 -- Tim Starling <tstarling@wikimedia.org>  Thu, 28 Apr 2016 16:55:25 +1000

html5depurate (1.0.1) unstable; urgency=medium

  * Fix trusty compilation error

 -- Tim Starling <tstarling@wikimedia.org>  Thu, 28 Apr 2016 16:03:33 +1000

html5depurate (1.0.0) unstable; urgency=low

  * Initial release

 -- Tim Starling <tstarling@wikimedia.org>  Thu, 14 Apr 2016 12:11:07 +1000


src/main/java/org/wikimedia/html5depurate/DepurateDaemon.java
package org.wikimedia.html5depurate;

import org.wikimedia.html5depurate.Config;
import org.wikimedia.html5depurate.DepurateErrorPageGenerator;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import org.apache.commons.daemon.Daemon;
import org.apache.commons.daemon.DaemonContext;
import org.apache.commons.daemon.DaemonInitException;

import org.glassfish.grizzly.http.server.HttpServer;
import org.glassfish.grizzly.http.server.NetworkListener;
import org.glassfish.grizzly.http.server.ServerConfiguration;

import java.io.FileNotFoundException;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.Properties;
import java.util.logging.Logger;
import java.util.logging.Level;

/**
 * Daemon for execution via jsvc
 */
public class DepurateDaemon implements Daemon {
	HttpServer m_server;
	String[] m_args;
	Logger m_logger = Logger.getLogger(this.getClass().getName());

	public static void main(String[] args) throws Exception {
		DepurateDaemon daemon = new DepurateDaemon();
		daemon.m_args = args;
		daemon.start();
		Thread.currentThread().join();
	}

	public void init(DaemonContext context)
			throws DaemonInitException,Exception
	{
		m_args = context.getArguments();
	}

	protected CommandLine loadCommandLine()
			throws IOException, ParseException
	{
		Options options = new Options();
		options.addOption("c", true, "The configuration file name");
		DefaultParser parser = new DefaultParser();
		return parser.parse(options, m_args);
	}

	protected Config loadConfig(String path) {
		Config config = new Config();
		Properties properties = new Properties();
		try {
			properties.load(new FileInputStream(path));
		} catch (FileNotFoundException e) {
			m_logger.warning("Config file not found: " + path);
		} catch (IOException e) {
			m_logger.warning("Error loading config file: " + e.toString());
		}

		double maxPostSize = Double.parseDouble(
				properties.getProperty("maxPostSize", "100e6"));
		if (maxPostSize > Integer.MAX_VALUE) {
			config.maxPostSize = Integer.MAX_VALUE;
		} else if (maxPostSize > 0) {
			config.maxPostSize = (int)maxPostSize;
		} else {
			config.maxPostSize = 100000000;
		}
		m_logger.info("Max post size: " + config.maxPostSize);

		config.host = properties.getProperty("host", "localhost");
		config.port = Integer.parseInt(properties.getProperty("port", "4339"));
		m_logger.info("Binding to " + config.host + ":" + Integer.toString(config.port));

		return config;
	}

	public void start() throws Exception {
		m_logger.info("Starting");

		CommandLine cl = loadCommandLine();
		String configPath = "/etc/html5depurate/html5depurate.conf";
		if (cl.hasOption("c")) {
			configPath = cl.getOptionValue("c");
		}
		Config config = loadConfig(configPath);

		m_server = new HttpServer();
		m_server.addListener(
				new NetworkListener("depurate", config.host, config.port));

		ServerConfiguration serverConf = m_server.getServerConfiguration();
		serverConf.addHttpHandler(new DepurateHandler(config));
		serverConf.setDefaultErrorPageGenerator(new DepurateErrorPageGenerator());
		serverConf.setName("depurate");
		m_server.start();
	}

	public void stop() throws Exception {
		m_logger.info("Stopping");
		if (m_server != null) {
			m_server.shutdownNow();
		}
	}

	public void destroy() {
	}
}
